### DISCUSSION TITLE
MQTT / Homeassistant

### DISCUSSION THREAD

**Discussion by lynntryanderror** [tbnobody/OpenDTU#156] [Q&A]
Hallo,

erstmal vielen Dank für dieses Projekt. Sehr coole Sache und hat gut funktioniert zu installieren.

Ich wollte jetzt MQTT einschalten damit ich die Werte in Homeassistant angezeigt bekommen und verwenden kann. Leider speichert er mir die Einstellungen nicht, Muss ich wenn ich "Enable Home Assistant MQTT Auto Discovery" aktiviere noch andere Parameter angeben? Wie zum Beispiel die MQTT Broker Parameter bzw. muss ich in HA nen MQTT am laufen haben?
Was genau ist der "Prefix Topic:"
Danke und viele Grüße
Lynn

**Comment by battifreund** [comment]
Hallo,

MQTT ist eine sog. Middleware, konkret ein Message Broker. Er verteilt Nachrichten von Sendern an Empfänger, die sog. Topics abonniert haben. Topics sind ähnlich hierarchisch strukturiert wie ein Dateisystem.

Also: Du brauchst einen MQTT-Broker. Den gibt es zB als AddOn für Home Assistant.

Hier die Anleitung zur Einrichtung: https://www.home-assistant.io/integrations/mqtt/

Im Falle von HA ist die ganze Sache ziemlich schlau gemacht. HA beobachtet in der Standardeinstellung den Topic /homeassistant.
In diesen Topic schreiben jetzt Devices eine Art Ankündigung, wie sie welche Daten an HA übermitteln möchten und in welchen Topics sich die Daten befinden. Das ist die AutoDiscovery-Funktion von HA.

Die eigentlichen Daten des Devices werden in eigenen Topics abgelegt. Bei der DTU ist es zB der Topic /solar.
Da schreibt die DTU die Messdaten der Wechselrichter rein und HA holt sie sich ab.

Zukünftig wird wahrscheinlich auch der umgekehrte Weg genutzt, um Einstellungen in der DTU vorzunehmen.

Du musst also, nachdem Du Deinen MQTT-Broker installiert und HA konfiguriert hast, nur noch folgende Einstellungen machen:

1. Enable MqTT -> ein
2. Enable Home Assistant MQTT Auto Discovery -> ein
3. Hostname -> Adresse des MQTT-Brokers
4. Port -> in der Regel 1883

Das reicht in der Regel schon. Ggf. musst Du noch den MQTT-Nutzer angeben, wenn Du ihn vorher entsprechend konfiguriert hast. Aber wenn ich mich richtig erinnere, ist der in der Standardkonfiguration des MQTT-AddOns nicht gesetzt.

Jetzt sollten die Daten in HA-Entities auftauchen.

**Reply by lynntryanderror** [reply]
Hi,

danke dir für die ausführliche Erklärung. Bisher habe ich noch nicht mit MQTT gemacht. Ich habe ein Gerät mit esphome zum laufen gebracht.

Ich hab jetzt den "Mosquitto broker" als add-in auf HA installiert. Bei der Configuration bin ich mir aber nicht sicher.
Was muss ich denn als Broker für einen namen eingeben? Eingetragen war "core-mosquitto"

Ich hätte jetzt erwartet beim HA server ein zusätzlicher Port geöffnet wird. Dem ist aber nicht so und HA findet auch nix neues.

![image](https://user-images.githubusercontent.com/101171559/191452663-ef471da0-9db2-4bf2-93e0-51409fbdee8c.png)

**Reply by battifreund** [reply]
Normalerweise ist es so, dass zwei Dinge passieren, wenn Du das AddOn installierst:
1. Ein Container mit Mosquitto wird gestartet und die Ports werden exportiert. 
Wenn also Deine HA-Instanz unter 192.168.1.42 läuft, erreichst Du Mosquitto unter mqtt://192.168.1.42:1883 (ohne TLS).
2. Die MQTT-Integration wird passend konfiguriert.
Daran musst Du nichts mehr ändern. Das scheint bei Dir auch passiert zu sein, sonst würdest Du den obigen Dialog nicht aufrufen können.

Es ist also so wie Du sagts: Mosquitto öffnet den Port auf Deiner HA-Instanz (wenn HA-OS verwendet wird).
Ich habe nachgeschaut und es ist so, dass Du einen User zur Anmeldung brauchst. Da geht jeder HA-User mit seinem Passwort. Alternativ könntest Du zum Test den Nutzer "homeassistant" mit dem oben zu sehenden (bzw. sichtbar zu machenden) Passwort verwenden.

Prinzipiell hast Du bis hierher alles richtig gemacht.

**Reply by lynntryanderror** [reply]
Also irgendwie macht das Mosquito MQTT Add-on nicht den Port 1883 auf. Ich muss mich am Wochenende mal hinsetzen und schauen ob ich da irgendwas finde was das Problem sein könnte. 

Danke schon mal für die Hilfe.

**Reply by lynntryanderror** [reply]
Ok ich bin ein wenig weiter. Warum auch immer konnte ich beim Port Scan vom HA Server den offenen Port nicht finden.
Im Log (ja da hätte ich früher drauf kommen können rein zu schauen...) sehe ich das.

2022-09-22 17:13:24: New connection from 192.168.1.72:60293 on port 1883.
2022-09-22 17:13:24: Client <unknown> disconnected, not authorised.

Dafür connected sich irgendwas anderes aus dem Netz...
2022-09-22 17:12:54: New connection from 172.30.32.1:59509 on port 1883.
2022-09-22 17:12:54: New client connected from 172.30.32.1:59509 as 28GRmcVB9hsmAVVnDXWqC8 (p2, c1, k60, u'homeassistant')

Hmmm

**Reply by battifreund** [reply]
> Dafür connected sich irgendwas anderes aus dem Netz... 2022-09-22 17:12:54: New connection from 172.30.32.1:59509 on port 1883. 2022-09-22 17:12:54: New client connected from 172.30.32.1:59509 as 28GRmcVB9hsmAVVnDXWqC8 (p2, c1, k60, u'homeassistant')

Das ist die HA-Integration, die sich vom anderen Container aus über das Docker-Network verbindet. Das ist also vollkommen korrekt.

Benutze mal als MQTT-User/Passwort den Nutzer, mit dem Du Dich an HA anmeldest. Mosquitto ist so konfiguriert, dass er gegen HA authentifiziert. Also kann man sich mit jedem HA-User am Mosquitto anmelden.

**Reply by lynntryanderror** [reply]
Danke dir für die Info das sich der HA drauf verbindet. Hatte mir schon Gedanken gemacht 🤪
Hab's jetzt zum laufen bekommen. Hab einen extra User/Passwort im Mosquito konfiguriert. Ich hatte beim OpenDTU am Anfang auch den User und Passwort genommen womit man sich auf dem Mosquito anmeldet. Jetzt Blick ich da schon mehr durch. Danke für eure Hilfe. Die nächste Frage folgt demnächst 😊

**Reply by realfogman** [reply]
Hallo,

habe nur eine kurze Frage zu MQTT / Home Assistant.

Werden bei euch alle Daten im MQTT Broker bzw. Home Assistant angezeigt?
Ich erhalte zwar Werte für Strom/Spannung/Leistung/Temperatur/etc. (auch der einzelnen Kanäle), allerdings nicht für "Yield Total", "Yield Day" und andere.
Unter der OpenDTU-Oberfläche werden diese Werte angezeigt und ich habe auch "HA MQTT Auto Discovery" eingestellt.
(Daher wurden mir diese Entitäten auch direkt unter HA vorgeschlagen/dargestellt.)

Sind hier die Sensor-Bezeichnungen evtl. verändert worden? (Es gibt z.B. / freq <- keine Werte /  und  / frequency <- Werte vorhanden / .)

Evtl. sollte ich auch ein neues Topic/Issue aufmachen, falls der Fehler nicht bei meiner Installation liegt?

**Reply by softwm8818** [reply]
also bei mir ist wieder alle Daten (auch "Yield Total", "Yield Day") da.  da hat sich als nichts geändert. 
schau doch mal unter INFO->MQTT ob du verbunden bist usw
viel erfolg

**Reply by tbnobody** [reply]
An der MQTT Ausgabe hat sich nichts geändert. Topics usw. sind alle identisch. Was siehst du denn für Werte im Broker selbst? Also einfach mal mit einem Tool wie MQTT-Explorer o.ä. verbinden und die Werte anschauen

**Reply by realfogman** [reply]
Hallo,
alle Werte sind auf grün (bis auf die Transportverschlüsselung TLS).
Ich gehe daher von einer vorhandenen Verbindung aus. (Ich bekomme ja einen Teil der Werte angezeigt.)
Leider nur nicht YieldTotal etc.

Ich werde mal versuchen die Firmware neu zu flashen und dem Gerät einen anderen Namen geben.

P.s. OpenDTU Firmware Version ist übrigens 0.1.21/ SDK Version v4.4.2.

**Reply by tbnobody** [reply]
Entscheident ist der Git-Hash. Das mit der Firmware Version muss ich mal umbenennen... 
Wie gesagt, MQTT-Explorer verwenden und schauen was direkt via MQTT ankommt.

**Reply by realfogman** [reply]
Hallo,

ich sehe gerade, dass die Topics unterschiedlich sind (HA -> MQTT -> HM Gerät -> MQTT Info (Debug-Informationen) ):

homeassistant/sensor/dtu_***********/ch0_PowerDC/config  (<- Wird angezeigt, Pakete werden regelmäßig empfangen)
versus:
homeassistant/sensor/HM-1500/ch0_YieldDay/config (<- Wird nicht angezeigt, keine Pakete erhalten)

Da die Topic Informationen m.M.n. durch OpenDTU an HM/MQTT versendet werden (auto discovery), dürfte hier ein Fehler vorliegen.

**Reply by realfogman** [reply]
So, jetzt läuft es!
Danke für den Tipp mit MQTT Explorer: Ist ein schönes Programm und auch gut als AddOn bei Home Assistant.

Zum Fehler:
Irgendwie hatte sich durch das ganze herumspielen meine Einstellungen beim MQTT Broker unter HA "verstellt".
Nachdem ich bemerkt habe, dass die Topics nicht stimmen, habe ich meine HM1500 aus dem Broker gelöscht, das Auto Discovery unter OpenDTU ausgeschaltet und einen Moment gewartet.
Danach wieder Auto Discovery eingeschaltet und die Veränderungen unter dem MQTT Broker im HA abgewartet.

Danke für die Hilfe!

**Comment by B00gieman** [comment]
I've tried all and give up...😞

Im using last HA and mosquitto broker.
I cant figure out how to use MQTT. 
I've reconfigured MQTT and cant connect.
Can anyone explain step by step how to?
Home Assistant side config and OpenDtu side config.

Thanks in advance

**Reply by realfogman** [reply]
Hello B00gieman,

first of all, there are a few things to check, before we can further evaluate the integration problem.
1. Is your OpenDTU device running and is it connected to your network (via wifi)?
2. Can you ping the IP address of your OpenDTU device, which your wifi (or dhcp server) has given your device?
3. Can you open the web interface of the OpenDTU device in your browser and have you already configured your Hoymiles device(s)? Do you receive information of your Hoymiles devices?

If you can answer all of the above with a "yes", then we can investigate further (in Home Assistant).
1. Can you reach your OpenDTU device from within your home assistant installation? (For example, via ping from the terminal addon?)
2. Is MqTT Broker installed and running as an addon in Home Assistant?

If so:
1. Check if you have enabled (on your OpenDTU device) Settings->Mqtt-Devices->"Enable MqTT" and "Enable Home Assistant MQTT Auto Discovery".
2. Check your "MqTT Broker Parameter" in OpenDTU (Settings->Mqtt-Devices).
3. Especially, recheck if you have the right IP-address, port-number and password (needs to be the same as in your Mqtt Broker installation in HA).

If all of the above is correct and working, then install the MQTT Explorer in Home Assistant!
(It seems to be a great tool!)
You will get an overview of the incomming/outgoing mqtt-messages.
If you receive the OpenDTU messages (sensor/ , solar/, dtu/status, etc. ???) then, at least, your OpenDTU device is sending the Hoymiles information towards Home Assistant.
If not, sometimes it helps to re-flash the OpenDTU device via OTA.

Greetings

**Reply by B00gieman** [reply]
Hello.
Thanks for your Help.
The answer is yes to all of your points ( i can't do ping, but that is other question...).

I've got the Mosquito broker as addon and configured. 
I think the problema must be on configs.
This os my log from Mosquitto Broker:

[](2022-10-13 14:22:48: New client connected from 192.168.1.106:59969 as OpenDTU-172028 (p2, c1, k15, u'nuno'). 2022-10-13 14:22:48: Invalid subscription string from 192.168.1.106, disconnecting. 2022-10-13 14:22:48: Client OpenDTU-172028 disconnected due to malformed packet.)

MQTT broker config
![Screenshot_2022-10-13-14-30-25-398_io.homeassistant.companion.android.jpg](https://user-images.githubusercontent.com/35974258/195611777-26c3d2fd-a0d5-43b7-b3cf-1de83a9e6346.jpg)
![Screenshot_2022-10-13-14-31-00-983_io.homeassistant.companion.android.jpg](https://user-images.githubusercontent.com/35974258/195611872-ed1a61c6-c54c-44cb-b3fb-2abc47d7bc65.jpg)

OpenDtu Mqtt config
![Screenshot_2022-10-13-14-32-18-610_com.brave.browser.jpg](https://user-images.githubusercontent.com/35974258/195612240-1c762ab6-cf1e-4f1f-a972-dd2fd0f3c54e.jpg)
![Screenshot_2022-10-13-14-32-29-188_com.brave.browser.jpg](https://user-images.githubusercontent.com/35974258/195612278-d18b2fcf-29c7-47a0-8d9d-bfa4119b1f38.jpg)
![Screenshot_2022-10-13-14-32-42-346_com.brave.browser.jpg](https://user-images.githubusercontent.com/35974258/195612324-fc3a8d44-41ad-4eb0-ae7f-e2a8540c3001.jpg)

**Reply by realfogman** [reply]
Hello B00gieman, just a quick question: Do you use some non-ASCII characters in your username and password?
Otherwise, your configurations seems to be alright?!

**Reply by B00gieman** [reply]
Hello.
No.
Username as you see "nuno" and password is simple 4 digit code.
Tomorrow i'll try to reinstall firmware from scratch.

**Reply by tbnobody** [reply]
Why did you change the default values for home assistant auto discovery topic? did you change this in your homeassistant installation as well? If not, just leave it as it is... ("homeassistant/", the trailing slash is important)

**Reply by tbnobody** [reply]
your base topic requires a trailing slash... and it should NOT be in the same topic as home assistant auto discovery. so leave it e.g. as "solar/"

**Reply by B00gieman** [reply]
Hello.
I confess that i dont know why i changed the configs...
As you sugested, and using my previous config i've changed DTU mqtt parameters to: 
Auto Discovery topic to homeassistant/
Base topic to solar/
(Slash is mandatory)

And..."voila" WORKING.
Thanks 
 

![Screenshot_2022-10-15-09-48-13-237_com.brave.browser.jpg](https://user-images.githubusercontent.com/35974258/195978144-5acedca0-3689-4681-a9ae-1ea27528c1eb.jpg)
![Screenshot_2022-10-15-09-47-55-958_io.homeassistant.companion.android.jpg](https://user-images.githubusercontent.com/35974258/195978152-d4b36f0b-04bf-433d-b567-8f1165dc0d2c.jpg)
![Screenshot_2022-10-15-09-55-31-733_io.homeassistant.companion.android.jpg](https://user-images.githubusercontent.com/35974258/195978193-48e664b4-f2d2-4b5e-9576-916581d2b425.jpg)

**Comment by kmarek1** [comment]
I have a problem with the Home Assistant Core2023.6.3 it does not detect the OpenDTU-A763B0 device. 
However, the inverter detects it without any problems.
In Moscito Broker you can see the device.
Might anyone have an idea how to fix this?


Log Mosquitto broker Current version: 6.2.1
``
2023-07-06 16:04:25: New connection from 192.168.50.50:58155 on port 1883.
2023-07-06 16:04:25: New client connected from 192.168.50.50:58155 as OpenDTU-10970032 (p2, c1, k15, u'homeassistant').

``
![scr1](https://github.com/tbnobody/OpenDTU/assets/117018507/d1a2b654-d44d-44d0-ab5f-741bd240012c)
![scr2](https://github.com/tbnobody/OpenDTU/assets/117018507/113465b1-2720-4543-8ad2-f8699b33e291)
![scr3](https://github.com/tbnobody/OpenDTU/assets/117018507/b214d253-22d9-4de3-8aa0-9eb402a84f7a)

**Comment by kmarek1** [comment]
I read it
https://www.home-assistant.io/integrations/sensor.mqtt/
https://github.com/lumapu/ahoy/issues/938
  and i did it like this
configuration.yaml: (add line)

`mqtt: !include mqtt.yaml # opendtu`

mqtt.yaml: (add new file in same directory as configuration.yaml if not existing yet)

```
sensor:
    # General data
  - name: "OpenDTU"
    icon: mdi:information-outline
    state_topic: "solar/dtu/status"
  - name: "OpenDTU power"
    device_class: power
    icon: mdi:solar-power
    unit_of_measurement: "W"
    state_topic: "solar/ac/power"  
  - name: "OpenDTU Cała produkcja"
    state_class: total
    device_class: energy
    icon: mdi:solar-power-variant
    unit_of_measurement: "kWh"
    state_topic: "solar/ac/yieldtotal"
  - name: "OpenDTU Dzienna produkcja"
    state_class: total
    device_class: energy
    icon: mdi:solar-power-variant
    unit_of_measurement: "Wh"
    state_topic: "solar/ac/yieldday"
```
![opendtu](https://github.com/tbnobody/OpenDTU/assets/117018507/1fef5a8c-628a-44a3-9f5e-30f536dbdf64)

**Reply by mangri01** [reply]
Hello, I followed your instructions.
The sensors are displayed in the entities in Home Assistant, but as unknown.
It's the same for all sensors.

In MQTT Explorer, everything from the openDTU is displayed, and all values ​​are live.

![Bildschirmfoto 2025-03-29 um 05 31 14](https://github.com/user-attachments/assets/8fd9a549-9fd0-4022-ac60-06fcfa5f61de)
![Bildschirmfoto 2025-03-29 um 05 31 36](https://github.com/user-attachments/assets/7c011528-bb96-4ee6-898f-5760bc2a0ebc)
![Bildschirmfoto 2025-03-29 um 05 31 54](https://github.com/user-attachments/assets/d17349e7-9e53-442a-8ce9-6f2f5b45e3d0)

**Comment by kmarek1** [comment]
Do you have Mosquitto broker installed and configured?

**Reply by mangri01** [reply]
Hello, I've installed the Mosquitto Broker. How should I configure it?
I think it's because I've already connected my Zendure system to MQTT.
<img width="511" alt="Bildschirmfoto 2025-03-30 um 10 45 50" src="https://github.com/user-attachments/assets/6b3e00d4-208d-4a57-85a0-1d6c35edc806" />



<img width="511" alt="Bildschirmfoto 2025-03-30 um 10 32 51" src="https://github.com/user-attachments/assets/a2f81791-797b-4cc4-81fa-e843edccca97" />
